org.eclipse.lsp4xml.extensions.maven/src/main/java/org/eclipse/lsp4xml/extensions/maven/MavenCompletionParticipant.java

```diff
@@ -113,11 +121,46 @@ public void onXMLContent(ICompletionRequest request, ICompletionResponse respons
 					phase -> phase.description, request).forEach(response::addCompletionAttribute);
 			break;
 		case "groupId":
-			if (!parent.getParentElement().getLocalName().equals("parent") && !parent.getParentElement().getLocalName().equals("plugin")) {
-				collectSimpleCompletionItems(localRepositorySearcher.searchGroupIds(RepositorySystem.defaultUserLocalRepository),
+			if (isParentDeclaration) {
+				// TODO local
+				collectParentCompletion(request, response);
+			} else {
+				// TODO if artifactId is set and match existing content, suggest only matching groupId
+				collectSimpleCompletionItems(isPlugin ? localRepositorySearcher.searchPluginGroupIds() : localRepositorySearcher.searchGroupIds(),
 						Function.identity(), Function.identity(), request).forEach(response::addCompletionAttribute);
+				internalCollectRemoteGAVCompletion(request, isPlugin).forEach(response::addCompletionItem);
+			}
+			break;
+		case "artifactId":
+			if (isParentDeclaration) {
+				// TODO local
+				collectParentCompletion(request, response);
+			} else {
+				(isPlugin ? localRepositorySearcher.getLocalPluginArtifacts() : localRepositorySearcher.getLocalArtifactsLastVersion()).keySet().stream()
+					.filter(groupIdArtifactId -> groupId.isEmpty() || groupIdArtifactId.groupId.equals(groupId.get()))
+					.map(groupIdartifactId -> groupIdartifactId.artifactId)
+					.distinct()
+					// TODO pass description as documentation
+					// TODO complete groupId if missing
+					// TODO? complete version
+					.map(artifact -> toCompletionItem(artifact, null, request.getReplaceRange()))
+					.forEach(response::addCompletionItem);
+				internalCollectRemoteGAVCompletion(request, isPlugin).forEach(response::addCompletionItem);
 			}
 			break;
+		case "version":
+			if (!isParentDeclaration) {
+				if (!artifactId.isEmpty()) {
+					localRepositorySearcher.getLocalArtifactsLastVersion().entrySet().stream()
+						.filter(entry -> entry.getKey().artifactId.equals(artifactId.get()))
+						.filter(entry -> groupId.isEmpty() || entry.getKey().groupId.equals(groupId.get()))
```

Comment:
Since we're using `Optional<String>`, we should do something like:
`.filter(entry -> (groupId.isPresent() && groupId.get().isEmpty()) || entry.getKey().groupId.equals(groupId.get()))`

Reply:
Same as above, we're sure that this optional would only contain a non-empty string.